packages/sqlite_async/lib/src/update_notification.dart

```diff
@@ -67,62 +70,118 @@ class UpdateNotification {
   }
 }
 
-/// Given a broadcast stream, return a singular throttled stream that is throttled.
-/// This immediately starts listening.
+/// Throttles an [input] stream to not emit events more often than with a
+/// frequency of 1/[timeout].
 ///
-/// Behaviour:
-///   If there was no event in "timeout", and one comes in, it is pushed immediately.
-///   Otherwise, we wait until the timeout is over.
-Stream<T> _throttleStream<T extends Object>(Stream<T> input, Duration timeout,
-    {bool throttleFirst = false, T Function(T, T)? add, T? addOne}) async* {
-  var nextPing = Completer<void>();
-  var done = false;
-  T? lastData;
-
-  var listener = input.listen((data) {
-    if (lastData != null && add != null) {
-      lastData = add(lastData!, data);
-    } else {
-      lastData = data;
+/// When an event is received and no timeout window is active, it is forwarded
+/// downstream and a timeout window is started. For events received within a
+/// timeout window, [add] is called to fold events. Then when the window
+/// expires, pending events are emitted.
+/// The subscription to the [input] stream is never paused.
+///
+/// When the returned stream is paused, an active timeout window is reset and
+/// restarts after the stream is resumed.
+///
+/// If [addOne] is not null, that event will always be added when the stream is
+/// subscribed to.
+/// When [throttleFirst] is true, a timeout window begins immediately after
+/// listening (so that the first event, apart from [addOne], is emitted no
+/// earlier than after [timeout]).
+Stream<T> _throttleStream<T extends Object>({
+  required Stream<T> input,
+  required Duration timeout,
+  required bool throttleFirst,
+  required T Function(T, T) add,
+  required T? addOne,
+}) {
+  return Stream.multi((listener) {
+    T? pendingData;
+    Timer? activeTimeoutWindow;
+    var needsTimeoutWindowAfterResume = false;
+
+    /// Add pending data, bypassing the active timeout window.
+    ///
+    /// This is used to forward error and done events immediately.
+    bool addPendingEvents() {
+      if (pendingData case final data?) {
+        pendingData = null;
+        listener.addSync(data);
+        activeTimeoutWindow?.cancel();
+        activeTimeoutWindow = null;
+        return true;
+      } else {
+        return false;
+      }
     }
-    if (!nextPing.isCompleted) {
-      nextPing.complete();
+
+    /// Emits [pendingData] if no timeout window is active, and then starts a
+    /// timeout window if necessary.
+    void maybeEmit() {
+      if (activeTimeoutWindow == null && !listener.isPaused) {
+        final didAdd = addPendingEvents();
+        if (didAdd) {
+          if (listener.isPaused) {
+            needsTimeoutWindowAfterResume = true;
+          } else {
+            activeTimeoutWindow = Timer(timeout, () {
+              activeTimeoutWindow = null;
+              maybeEmit();
+            });
```

Comment:
Could this re-use the `setTimeout()` function below?

Reply:
In Dart we can't call local functions before they're declared (so the scoping is a bit different than in say JS), but I've turned this into a late function to reuse it.

Reply:
Ah, I see how that makes it more difficult with the circular dependency between the two functions.